src/DetailsView/components/failure-instance-panel-control.tsx

```diff
@@ -133,39 +150,73 @@ export class FailureInstancePanelControl extends React.Component<FailureInstance
     private getFailureInstancePanelDetails = (): JSX.Element => {
         return (
             <FailureInstancePanelDetails
-                path={this.state.path}
-                snippet={this.state.snippet}
+                path={this.state.currentInstance.path}
+                snippet={this.state.currentInstance.snippet}
                 onSelectorChange={this.onSelectorChange}
                 onValidateSelector={this.onValidateSelector}
             />
         );
     };
 
+    private getCurrentInstance = (): FailureInstanceData => {
+        const updatedInstance = {
+            failureDescription: this.state.currentInstance.failureDescription,
+            path: this.state.currentInstance.path,
+            snippet: this.state.currentInstance.snippet,
+        };
+
+        return updatedInstance;
+    };
+
     protected onFailureDescriptionChange = (event, value: string): void => {
-        this.setState({ failureDescription: value });
+        const updatedInstance = this.getCurrentInstance();
+        updatedInstance.failureDescription = value;
+        this.setState({ currentInstance: updatedInstance });
     };
 
     private onSelectorChange = (event, value: string): void => {
-        this.setState({ path: value });
+        const updatedInstance = this.getCurrentInstance();
+        updatedInstance.path = value;
+        this.setState({ currentInstance: updatedInstance });
     };
 
     private onValidateSelector = (event): void => {
-        const currSelector = this.state.path;
-        this.setState({ snippet: 'snippet for ' + currSelector });
+        const currSelector = this.state.currentInstance.path;
+        const currSnippet = 'snippet for ' + currSelector;
+        const updatedInstance = this.getCurrentInstance();
+        updatedInstance.snippet = currSnippet;
+        this.setState({ currentInstance: updatedInstance });
     };
 
     protected onAddFailureInstance = (): void => {
-        this.props.addFailureInstance(this.state.failureDescription, this.props.test, this.props.step);
+        this.props.addFailureInstance(this.state.currentInstance, this.props.test, this.props.step);
         this.setState({ isPanelOpen: false });
     };
 
     protected onSaveEditedFailureInstance = (): void => {
-        this.props.editFailureInstance(this.state.failureDescription, this.props.test, this.props.step, this.props.instanceId);
+        this.props.editFailureInstance(this.state.currentInstance, this.props.test, this.props.step, this.props.instanceId);
         this.setState({ isPanelOpen: false });
     };
 
     protected openFailureInstancePanel = (): void => {
-        this.setState({ isPanelOpen: true, failureDescription: this.props.originalText || '' });
+        let propsDescription = '';
```

Comment:
This function seems a bit convoluted to me.

You are currently creating properties and then structuring them to be an object. You can actually just create a default object and then just combine the instance from the props.

```
const defaultInstance = getDefaultInstance();
const currentInstance = {
    ...defaultInstance,
    ...this.props.originalInstance,
}
this.setState({
            isPanelOpen: true,
            currentInstance: currentInstance,
});
```

**Some of this** could also simplify your constructor.

Reply:
Great! This definitely cleaned things up--worked in the constructor as well!